[no squash] Add animated glTF support

appgurueu opened the pull request:
This is the follow-up to https://github.com/minetest/minetest/pull/14557. Closes #9673.

## How to test

Test entities and unit tests are included.

appgurueu commented:
Rebased. The static PR is ready. This is ready too.

appgurueu commented:
To clarify the state of this PR: This PR is ready for review, as is the static PR. It depends on the static PR however, hence the "waiting on dependency" tag.

rubenwardy commented on doc/lua_api.md:
```diff
@@ -304,7 +304,10 @@ it unlocks no special rendering features.
+* Animations
+  * Only a single animation is supported,
+    use frame ranges within this animation.
+  * Only integer frames are supported.
```
how easy is it to fix? Sounds like something needed for useful gltf animation support

appgurueu replied:
https://github.com/minetest/minetest/pull/14685/commits/d9d6fcc23082f2e965ca1112cf4cc97b05636315

appgurueu commented:
I have squashed most commits to make this easier to review and merge. I recommend reviewing this by looking at the individual commits.

appgurueu commented on irr/include/matrix4.h:
```diff
@@ -24,7 +24,7 @@ namespace core
 
 //! 4x4 matrix. Mostly used as transformation matrix for 3d calculations.
-/** The matrix is a D3D style matrix, row major with translations in the 4th row. */
+/** The matrix is stored in column major order, with translations in the 4th row. */
```
Heh "row" here should be "column" of course.

appgurueu commented:
Per request of someone who tried to understand the matrix code and could have been helped by my refactor, I've split off https://github.com/minetest/minetest/pull/14685/commits/3446bd1001d876053899aabb00615f3ef081b803 into a separate PR: #15216 and rebased this one on it.

sfan5 commented on games/devtest/mods/gltf/models/gltf_simple_skin.gltf:
```diff
@@ -0,0 +1 @@
+{"scene":0,"scenes":[{"nodes":[0,1]}],"nodes":[{"skin":0,"mesh":0},{"children":[2]},{"translation":[0.0,1.0,0.0],"rotation":[0.0,0.0,0.0,1.0]}],"meshes":[{"primitives":[{"attributes":{"POSITION":1,"JOINTS_0":2,"WEIGHTS_0":3},"indices":0}]}],"skins":[{"inverseBindMatrices":4,"joints":[1,2]}],"animations":[{"channels":[{"sampler":0,"target":{"node":2,"path":"rotation"}}],"samplers":[{"input":5,"int ...
```
offtopic but we could use .gitattributes to have git treat .gltf like binary files (so they wouldn't show up in a text diff)

appgurueu replied:
included in the commit. doing a `git show` locally seems to respect it. it doesn't look like GitHub does however (at least not for this PR).

sfan5 commented on irr/include/IAnimatedMesh.h:
```diff
@@ -39,19 +36,9 @@ class IAnimatedMesh : public IMesh
-	virtual IMesh *getMesh(s32 frame, s32 detailLevel = 255, s32 startFrameLoop = -1, s32 endFrameLoop = -1) = 0;
+	/** \param frame: Frame number, >= 0, <= getMaxFrameNumber()
+	\return Returns the animated mesh for the given frame */
+	virtual IMesh *getMesh(f32 frame) = 0;
```
I would prefer a short comment saying what happens when I request frame 1.5 and only 1 and 2 exist.
or are float frames even allowed?

appgurueu replied:
It linearly interpolates. Added a comment.

sfan5 commented on irr/src/CGLTFMeshFileLoader.cpp:
```diff
@@ -385,48 +410,114 @@ static std::vector<u16> generateIndices(const std::size_t nVerts)
+{
+	m_mesh_loaders.emplace_back([=] {
+		for (std::size_t pi = 0; pi < getPrimitiveCount(meshIdx); ++pi) {
+			const auto &primitive = m_gltf_model.meshes->at(meshIdx).primitives.at(pi);
```
not important: I bet a few indentation levels can be saved here.

appgurueu replied:
you're right, there was a function to be extracted lurking here

sfan5 commented on irr/src/CGLTFMeshFileLoader.h:
```diff
@@ -114,12 +117,15 @@ class CGLTFMeshFileLoader : public IMeshLoader
 		const tiniergltf::GlTF m_gltf_model;
 		CSkinnedMesh *m_irr_model;
 
+		std::vector<std::function<void()>> m_mesh_loaders;
```
this only ever has exactly one element. is the abstraction worth it?

appgurueu replied:
In practice this will almost always have exactly one element, but in principle a gltf model can have multiple meshes, in which case it should still work correctly.

sfan5 commented on irr/include/vector2d.h:
```diff
@@ -34,6 +34,11 @@ class vector2d
+	template <class U>
+	constexpr vector2d<U> cast() const {
+		return vector2d<U>(static_cast<U>(X), static_cast<U>(Y));
+	}
```
:+1: for idea but maybe
`template <class U> constexpr static vector2d<T> from(const vector2d<U> &other)`
would make more sense conceptually.
`a.cast<f32>()` vs. `v3f::from(a)`

appgurueu replied:
done

sfan5 commented on src/network/networkprotocol.h:
```diff
@@ -229,14 +229,15 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 			old servers.
 		Rename TOCLIENT_DEATHSCREEN to TOCLIENT_DEATHSCREEN_LEGACY
 		Rename TOSERVER_RESPAWN to TOSERVER_RESPAWN_LEGACY
+		Support float animation frame numbers
```
```suggestion
		Support float animation frame numbers in TOCLIENT_LOCAL_PLAYER_ANIMATIONS
```

sfan5 commented:
> Per request of someone who tried to understand the matrix code and could have been helped by my refactor, I've split off [3446bd1](https://github.com/minetest/minetest/commit/3446bd1001d876053899aabb00615f3ef081b803) into a separate PR: #15216 and rebased this one on it.

I fear this PR may be ready earlier than the matrix one.

appgurueu commented:
> I fear this PR may be ready earlier than the matrix one.

I've now put the matrix PR in a state which I believe to be uncontroversial.

sfan5 commented on .gitattributes:
```diff
@@ -3,3 +3,5 @@
 *.cpp diff=cpp
 *.h diff=cpp
+
+*.gltf -diff
```
alternatively
```suggestion
*.gltf binary
```

appgurueu replied:
ah yeah, that does a few more things than just `-diff`

sfan5 reviewed (approved):
not tested but thankfully this has tests :)
